Add a 'skip' parameter to writev1 so that the beginning of a car can …

willscott opened the pull request:
…be skipped

dirkmc commented on v2/internal/loader/writing_loader.go:
```diff
@@ -50,17 +51,31 @@ type writingReader struct {
 }
 
 func (w *writingReader) Read(p []byte) (int, error) {
+	buf := bytes.NewBuffer(nil)
```
Is it possible to avoid creating a new buffer here unless the skip falls within this block?

willscott replied:
the complexity of getting the potential partial-write edge cases of each of the 3 internal writes worries me a little bit, but yeah i'll take a look

masih commented on v2/selective.go:
```diff
@@ -38,17 +40,36 @@ func MaxTraversalLinks(MaxTraversalLinks uint64) Option {
 }
 
+// WithV1Size sets the expected v1 size of the car being written if it is known in advance.
+func WithV1Size(size uint64) Option {
```
```suggestion
func WithDataPayloadSize(size uint64) Option {
```
The term `DataPayload` is used to describe the inner CARv1 in [CARv2 spec](https://ipld.io/specs/transport/car/carv2/#carv1-data-payload). 
It's also a term used across other v2 APIs.

masih commented on v2/selective.go:
```diff
@@ -101,19 +123,40 @@ func TraverseToFile(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, sele
 }
 
+// CreateV1Reader creates an io.ReadSeeker that can be used to copy out the carv1 contents of a car.
+func CreateV1Reader(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, opts ...Option) (io.ReadSeeker, error) {
```
* Use `New` instead of `Create` for consistency with existing APIs?
* This is a root level function. Considering it applies to the "selective" read/write context, can I ask you to rename this to something less generic.

willscott replied:
still waiting to figure out what the partial traversal API in ipld-prime will look like. this may be a function on a traversalCar instead, depending on where we want shared cache state between different traversals to live.

willscott replied:
I have now lost context here and am not sure if my argument about partial traversal holds

willscott commented:
The remaining potential work here is allow skip to be an arbitrary byte offset in the car file. currently it's limited to block boundaries, and because of the API it's difficult to know which block boundary nearest to the requested skip offset has been used.
If it's acceptable for the caller to skip only to block boundaries, this could be sufficient as-is, otherwise the wrapper for the individual reader needs to be updated to also partially skip.

willscott commented:
I cleaned up the tee-ing writer and added a test that seems to verify that we can use this to advance to arbitrary byte offsets in a car.

@rvagg maybe can provide a first round of review

rvagg commented on v2/internal/loader/counting_loader.go:
```diff
@@ -9,13 +9,13 @@ import (
-type counter struct {
-	totalRead uint64
+// Counter tracks how much data has been read.
+type Counter struct {
```
why do we use a struct for this? is it something to do with needing a pointer to it? why isn't a `*uint64` good enough for this purpose?

rvagg commented on v2/traversal/resumption.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func (pn pathNode) addPath(p []datamodel.PathSegment, link datamodel.Link, at uint64) {
+	if len(p) == 0 {
```
is this just for the root case where we've made a `newPath()` and don't want to add the root to itself?

rvagg commented on v2/traversal/resumption.go:
```diff
@@ -0,0 +1,257 @@
+
+// TraverseResumer allows resuming a progress from a previously encountered path in the selector.
+type TraverseResumer interface {
+	RewindToPath(from datamodel.Path) error
```
am I right in assuming we don't have a use-case for this at the moment, that we're just supporting it because it's not a difficult extension of doing it to an offset?

got thoughts on what a use-case for might me?

willscott replied:
This is probably the most useful thing to have pointed out. I think the more useful method to expose here will be a `RewindToOffset` in addition to the current `RewindToPath`. I'll work on adding that

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -46,25 +49,50 @@ type IndexTracker interface {
+	buf []byte
 	cid string
-	wo  *writerOutput
+	wo  *indexingWriter
```
maybe rename this so it's not got an unexplainable name for future eyes?

```suggestion
	iw  *indexingWriter
```

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -46,25 +49,50 @@ type IndexTracker interface {
+		if w.wo.toSkip > 0 {
+			if w.wo.toSkip >= uint64(len(writeBuf)) {
+				w.wo.toSkip -= uint64(len(writeBuf))
+				writeBuf = []byte{}
```
```suggestion
				// will cause the WriteTo() below to be a noop, we need to skip this entire block
				writeBuf = []byte{}
```

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -46,25 +49,50 @@ type IndexTracker interface {
 		}
-		cpy := bytes.NewBuffer(w.r.(*bytes.Buffer).Bytes())
-		if _, err := cpy.WriteTo(w.wo.w); err != nil {
+		sizeBytes := varint.ToUvarint(uint64(n) + uint64(len(w.cid)))
```
```suggestion
		// write the varint size prefix and trim the unneeded prefix padding we allocated
		sizeBytes := varint.ToUvarint(uint64(n) + uint64(len(w.cid)))
```

rvagg commented on v2/selective.go:
```diff
@@ -202,21 +250,38 @@ func (tc *traversalCar) WriteV2Header(w io.Writer) (int64, error) {
+		skip -= v1Size
 	}
 
 	// write the block.
```
```suggestion
	// write the blocks
```

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -75,11 +103,19 @@ func (w *writingReader) Read(p []byte) (int, error) {
 		w.wo = nil
 	}
 
+	if w.buf != nil {
```
```suggestion
	if w.buf != nil {
		// we've already read the block from the parent reader for writing the CAR block section (above),
		// so we need to pass those bytes on in whatever chunk size the caller wants
```

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -46,25 +49,50 @@ type IndexTracker interface {
 
+var _ IndexTracker = (*indexingWriter)(nil)
+
 type writingReader struct {
```
```suggestion
// writingReader is used on a per-block basis for the TeeingLinkSystem's StorageReadOpener, we use it
// to intercept block reads and construct CAR section output for that block, passing that section data to
// indexingWriter, while also passing the plain binary block data back to the LinkSystem caller (which
// we expect to be a traversal operation).
// Additionally, if we are performing a "skip" of initial bytes for this CAR, we track the byte count as we
// construct the CAR section data and decide when and how much to write out to the indexingWriter.
// Skip operations don't impact the downstream LinkSystem user (traversal), but they do impact what's
// written out via the indexingWriter.
type writingReader struct {
```

rvagg commented on v2/internal/loader/writing_loader.go:
```diff
@@ -46,25 +49,50 @@ type IndexTracker interface {
+		_ = copy(writeBuf[:], sizeBytes)
+
+		size := len(writeBuf)
+		if w.wo.toSkip > 0 {
```
```suggestion
		// indexingWriter manages state for a skip operation, but we have to mutate it here -
		// if there are still bytes to skip, then we either need to skip over this whole block, or pass
		// part of it on, and then update the toSkip state
		if w.wo.toSkip > 0 {
```

rvagg commented on v2/internal/io/skip_writer_read_seeker.go:
```diff
@@ -0,0 +1,108 @@
+var _ io.ReadSeeker = (*SkipWriterReaderSeeker)(nil)
+
+// NewSkipWriterReaderSeeker creates an io.ReadSeeker around a ReWriter.
+func NewSkipWriterReaderSeeker(ctx context.Context, size uint64, cons ReWriter) *SkipWriterReaderSeeker {
```
could we have a better name than `cons`? it's not obvious to me what this name is supposed to be (`consume`?), it's puzzling when we get to `c.cons()` and it seems like it should be more obvious what it's doing - should it be read as "start consuming now"?

rvagg commented on v2/selective.go:
```diff
@@ -103,19 +128,40 @@ func TraverseToFile(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, sele
+		opts:     conf,
+	}
+	rwf := func(ctx context.Context, offset uint64, writer io.Writer) (uint64, error) {
+		s, _, err := tc.WriteV1(ctx, offset, writer)
```
```suggestion
		// it's only at this point we have the `offset` to start writing at since the user of the
		// ReadSeeker will (may) have called Seek() and we've worked out where in the CAR
		// that is as an offset. Now we can start writing the CARv1 data, which will be passed
		// on to the ReadSeeker.
		// Note that we're inside a goroutine here
		s, _, err := tc.WriteV1(ctx, offset, writer)
```

rvagg commented on v2/internal/io/skip_writer_read_seeker.go:
```diff
@@ -0,0 +1,108 @@
+	// Check if there's already a write in progress
+	if c.reader == nil {
+		// No write in progress, start a new write from the current offset
+		// in a go routine
```
```suggestion
		// in a go routine and feed it back to the caller via a pipe
```

rvagg commented on v2/selective.go:
```diff
@@ -40,40 +43,62 @@ func MaxTraversalLinks(MaxTraversalLinks uint64) Option {
 	}
 }
 
+// WithDataPayloadSize sets the expected v1 size of the car being written if it is known in advance.
```
```suggestion
// WithDataPayloadSize sets the expected v1 size of the car being written if it is known in advance.
// This is required if NewCarV1StreamReader() is used and a Seek() operation needs seek back from
// SeekEnd (i.e. if we don't know where the end is, we can't figure out how far that is from the start).
```

rvagg commented on v2/selective.go:
```diff
@@ -103,19 +128,40 @@ func TraverseToFile(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, sele
 }
 
+// NewCarV1StreamReader creates an io.ReadSeeker that can be used to copy out the carv1 contents of a car.
+func NewCarV1StreamReader(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, opts ...Option) (io.ReadSeeker, error) {
```
this should probably suggest "Selective" shouldn't it? to match `NewSelectiveWriter()`. Perhaps:

```suggestion
func NewSelectiveV1Reader(ctx context.Context, ls *ipld.LinkSystem, root cid.Cid, selector ipld.Node, opts ...Option) (io.ReadSeeker, error) {
```

rvagg commented:
_(FYI I messed this branch up a tiny bit so ended up rebasing it on master, removing the merge commits in the process, so the commits have changed but the resulting code is the same)_

willscott commented:
@rvagg are we able to merge this - was there something else we were waiting for on it?

rvagg commented:
Yeah, it needs #327 to get closer to working as advertised, but even then it's still not quite right. I stepped back from working on this as the complexity increased and I started questioning the utility since we mainly want this (for now) for the explore-all case and it's the selectors that make this difficult to properly do.

As it is, it can't replace Boost's impl because of caching enabled by explore-all shortcuts that we're not allowing ourselves to touch here.

I'd be fine merging in #327 to here, but I'm not sure we should merge this in since it's not going to give people what it says it does.

rvagg commented:
I forgot to add that I'm not intending to suggest this is a lost cause. My thinking, the last time I touched this, was that we should try to implement an efficient explore-all case here, and then iterate on that for the more complex cases.
